### ISSUE TITLE
Build errors rpi3

### ISSUE THREAD

**Issue by johncblacker** [superuser789/MediaPipe-on-RaspberryPi#5]
I tried a build and got errors:
cannot find -lImath
cannot find -lIlmImf
cannot find -lIex
cannot find -lHalf
cannot find -lIlmThread
cannot find -ldc1394

Installed dependencies noted in other build error issue, installed ffmpeg and opencv from this repository.

**Comment by johncblacker** [comment]
Still trying...found a post that suggested modifying the third_party BUILD with NEON off.  Rebuilt and still got same error.  I'm going to uninstall the repo I built, re-install from the original mediapipe-rpi3 that I have and give my example a try.  If it works, then my build is the problem; if it doesn't work, then my environment must be the issue.  Will let you know.

**Comment by johncblacker** [comment]
The result of my experiment are:  your original build works, my build does not.  So, I guess I made some kind of mistake somewhere in the build.  Before I undertake another build (which on my RPi takes several hours) I'd like to find out definitively what my error is (or I could just run your latest build presuming you're using the latest mediapipe sources.  The latest sources support the use of the "model_selection=" parameter on the init of the FaceDetection class as documented on the mediapipe web page.

**Comment by johncblacker** [comment]
Tried copying the _framework...so file from your distro to my site-packages just to see if that'd make a difference and it did not.
So, now I'm re-doing the build but following the instructions from one of the mediapipe support folks.  This'll take a long time...hours...be done this evening sometime.

**Comment by johncblacker** [comment]
I give up!  I've tried everything I could think of to build the mediapipe from latest sources;  I even had problems that were generally diagnosed as bazel problems and had to go back and try to find and build a version of bazel that'd work.  All I ended up doing was going down more and more rabbit holes and burning up time and getting frustrated at the lack of general support for both mediapipe and bazel - especially on the RPi.  So, I'll fall back to your current build and move on.  If you ever get the latest sources to build, let me know.

**Comment by johncblacker** [comment]
Well, I haven't given up - yet.  I used nm -A to list the symbols in the _framework_bindings.cpython-37m-arm-linux-gnueabihf.so file in the mediapipe-rpi3 version (0.8.4) and in the version I built from current sources and the symbol that appears in the error as being undefined is the same in both of the files so I can't explain why it would be reported as an undefined symbol - IT'S THERE!
Have you made any progress building from current sources?

**Comment by superuser789** [comment]
> Decided to give a rebuild from the mediapipe sources and followed the directions in your README.md. Tried running a simple face detection app and got a failure with "import mediapipe as mp" statement. The output is:
> ******************************* output starts next line ******************************
> Traceback (most recent call last):
> File "", line 1, in
> File "/home/pi/ml-projects/ex-mediapipe/mymedpipeEnv/lib/python3.7/site-packages/mediapipe/**init**.py", line 16, in
> from mediapipe.python import *
> File "/home/pi/ml-projects/ex-mediapipe/mymedpipeEnv/lib/python3.7/site-packages/mediapipe/python/**init**.py", line 17, in
> from mediapipe.python._framework_bindings import resource_util
> ImportError: /home/pi/ml-projects/ex-mediapipe/mymedpipeEnv/lib/python3.7/site-packages/mediapipe/python/_framework_bindings.cpython-37m-arm-linux-gnueabihf.so: undefined symbol: _ZN3ruy28Pack8bitColMajorForNeon2ColsERKNS_14PackParams8bitE
> ********************* end of output ******************************************
> Don't know where to go from here, perhaps something I did or didn't do in the build? Do you have any ideas?
> Thanks.

The solution was posted by [arron2003](https://github.com/arron2003)  [here](https://github.com/google/mediapipe/issues/1629#issuecomment-814599336). Add those flags & rebuilt it again. 
I also noticed that there is some vagueness in [Step 8](https://github.com/superuser789/MediaPipe-on-RaspberryPi) as there are two lines with '--compilation_mode=opt'.  Those flags are needed to be added in function `bazel_build` inside `class BuildBazelExtension` after `'--compilation_mode=opt'`. Sorry for that.

**Comment by superuser789** [comment]
> Well, I haven't given up - yet. I used nm -A to list the symbols in the _framework_bindings.cpython-37m-arm-linux-gnueabihf.so file in the mediapipe-rpi3 version (0.8.4) and in the version I built from current sources and the symbol that appears in the error as being undefined is the same in both of the files so I can't explain why it would be reported as an undefined symbol - IT'S THERE!
> Have you made any progress building from current sources?

Apologies. I was caught up with some work.
I built the latest mediapipe release 0.8.7 & ran it successfully. You may test the latest package for RPi 3 from [here](https://github.com/superuser789/MediaPipe-on-RaspberryPi/raw/main/RPi%203/mediapipe-0.8.7-cp37-cp37m-linux_armv7l.whl).

**Comment by johncblacker** [comment]
Not a problem, I was busy with other projects as well.  Thanks for the re-build.  I'm building with a change suggested by mediapipe support.  Unfortunately, it's 4-5hr build.  After that build completes I'm going to give it a shot and if it fails I'll load your build and give it a shot.  I'm still curious how you can build successfully, while I can't.  Are you building on a pi? or are you building on a rooted phone?  If a phone, which one and how did you set it up - just curious.

**Comment by superuser789** [comment]
> Not a problem, I was busy with other projects as well. Thanks for the re-build. I'm building with a change suggested by mediapipe support. Unfortunately, it's 4-5hr build. After that build completes I'm going to give it a shot and if it fails I'll load your build and give it a shot. I'm still curious how you can build successfully, while I can't. Are you building on a pi? or are you building on a rooted phone? If a phone, which one and how did you set it up - just curious.

I first build on a rooted LG G8X having snapdragon 855 processor (8 cores) & 6GB LPDDR4X ram. I have mounted Raspbian system img and run in chroot environment using chroot script available for android.

**Comment by johncblacker** [comment]
I have more questions in my effort to understand why your builds work and mine don't.  If you could please help?
1.)  are you running a raspbian lite or desktop and is it basically a minimal install?
2.)  what version of bazel are you running?
3.)  what version (if any) of opencv did you install via apt (if any);
4.)  are you performing the build under root?
5.)  are you following the "recipe" posted in the readme for rPi 3?
6.)  when you test, do you create a "new" environment (presuming you're using one at all) or are you using the global one?
7.)  what's in your environment when you install the mediapipe wheel; or are you letting pip load the dependencies?
8.)  what version of opencv-contrib is in the python site-packages?

Thanks for your help.  I'm planning on grabbing a new sd card, reinstalling raspbian and trying the build with a minimal install because I think there just has to be something in my environment that is messing up the mediapipe build.  Your help has been much appreciated and answers to the above questions would go a long way towards helping me out.  BTW, your build worked on my machine (0.8.7 version) and it supports the use of "model_selection=0 (or 1)" depending on whether you want to use the short-range or long-range model.  THANKS!

**Comment by superuser789** [comment]
No problem.

1. raspbian lite 
2. bazel 4.0.0
3. OpenCV 4.5.2  (Not official ; built using [guide](https://github.com/superuser789/MediaPipe-on-RaspberryPi/blob/main/BuildingFFMPEG%26OpenCV.md) )
4. As user Pi
5. Yes, I build them using [guides](https://github.com/superuser789/MediaPipe-on-RaspberryPi#readme).
6. Just the global one
7. First I install the required dependency packages using pip. Then install the mediapipe package
8. Same version as that of OpenCV i.e. 4.5.2

Could you describe the problem you are facing while building ?

**Comment by johncblacker** [comment]
I was reading the requirements for the ffmpeg install tutorial and he specifically mentions using "stretch" - is that what you're using or have you found that you can safely use the latest "buster?"
I'm going to give the build another shot.  If need be I can pull out another sd card and load an earlier image to it.

**Comment by superuser789** [comment]
I am building on the latest Raspbian buster (10). All the packages are updated.

**Comment by johncblacker** [comment]
I tried building ffmpeg using the vidware script and ran into an error:

**Comment by superuser789** [comment]
You can try this [guide](https://github.com/superuser789/MediaPipe-on-RaspberryPi/blob/main/BuildingFFMPEG%26OpenCV.md#building-ffmpeg). Make sure that the packages fdk-aac, mp3lame, libass, x264 are already installed and remove any pre installed x264 package.
I run ssh server (in chroot raspbian) on my LG phone and run commands over ssh using putty ssh client  on windows.

**Comment by johncblacker** [comment]
I'm a little confused.  Your comment says to ensure the 4 packages listed are already installed, but then you say remove any pre-installed x264 package.  So, removed the x264 but leave the others installed?  BTW, I've tried with a vanilla raspbian-lite Buster image, but not everything completed successfully and the ffmpeg .deb package wasn't built nor did the "mpv" command test work either.  Don't know if there's any support on github for this, but there's the caveat about test with "stretch" only so I don't expect the developer to support on Buster.

**Comment by superuser789** [comment]
> I'm a little confused. Your comment says to ensure the 4 packages listed are already installed, but then you say remove any pre-installed x264 package. So, removed the x264 but leave the others installed? BTW, I've tried with a vanilla raspbian-lite Buster image, but not everything completed successfully and the ffmpeg .deb package wasn't built nor did the "mpv" command test work either. Don't know if there's any support on github for this, but there's the caveat about test with "stretch" only so I don't expect the developer to support on Buster.

A particular ffmpeg is built on specific versions of fdk-aac, mp3lame, libass, x264.

First remove the pre installed fdk-aac, mp3lame, libass, x264, ffmpeg packages & these also :
`sudo apt remove libavcodec-dev libavutil-dev libavformat-dev libswscale-dev libavutil56`

Either install from official repository : `sudo apt install ffmpeg`
or build all the packages using [guide](https://www.raspberrypi.org/forums/viewtopic.php?t=199775).

**Comment by superuser789** [comment]
> I read the guide and copied the script – that’s what I was using to build; however, I keep getting errors near the end, beginning with the build of ffmpeg and onward. I think I’m about done trying to build mediapipe – I’ve spent way too much time on it and have other things to do. I even read most of the dialog following the tutorial looking for an answer, but didn’t find one from RPi-Mike the producer of the script to build ffmpeg, etc. While I can understand he put a lot of work into it, I also believe that it’s going to become “stale” unless a solution for Buster and beyond becomes available. Thanks for your help. Your build of mediapipe works well and I was able to switch the model to the longer-range one in order to detect faces at a greater distance – which is exactly what I needed from the latest release – THANKS AGAIN! John

Thanks for trying out the mediapipe package.
I had successfully built the latest ffmpeg & mpv on raspbian buster.
Could you share the error logs you are getting while building ffmpeg, mpv and mediapipe ?

**Comment by johncblacker** [comment]
I have attached a txt file containing the build log from building mpv -on which is where the failure is now;  At the end you’ll see a list of the .deb files created and what’s missing is mpv.  No point in trying to build mediapipe until I can get this build script to complete, is there?  I know I can install ffmpeg and opencv using apt, but that might just put me back to my original mediapipe runtime error which is:

“ImportError: /home/pi/ml-projects/ex-mediapipe/mymedpipeEnv/lib/python3.7/site-packages/mediapipe/python/_framework_bindings.cpython-37m-arm-linux-gnueabihf.so: undefined symbol: _ZN3ruy28Pack8bitColMajorForNeon2ColsERKNS_14PackParams8bitE”

It appears from my research that a number of people are getting a similar error (although the symbol itself differs) and mediapipe support is blaming it on a problem in tf-lite.  I opened an issue on the mediapipe git, but the response was “there’s nothing we can do” and someone suggested I open a new issue regarding a request for new functionality, which I did (I asked them for full build support of mediapipe on Raspberry Pi).  I also relayed my dismay for the fact that mediapipe and tensor-flow and tf-lite are all Google creations; therefore, it’s appalling that the groups can’t communicate amongst themselves to fix the problem!  Now what’s interesting is that you’re able to build on the Raspberry Pi, whereas I’m not.  So, that begs the question about what is it that’s different about my environment (I’m presuming it’s software related) compared to yours that results in the error.  I’m using a recent version of Raspbian Desktop, running on a Raspberry Pi 3b+.  I do admit that I’ve been working on a number of projects relating to motion and face detection and have installed supporting software.  I also have a Google Coral USB device to support tf-lite – that’s my “normal environment” but I did recently create a Raspbian Buster lite image with nothing added to it but what seems to be required to support the vidware script (namely, checkinstall in order to create the .deb packages).  So, once I get this ffmpeg build to work (if I ever do) then I’ll move on to the opencv build, then the mediapipe build and see what happens with the “vanilla” Raspbian-lite environment.  Of course the mediapipe build on the rpi takes over 5 hours! 

Sorry about the long story, but I’ve spent weeks on this (with some time off to try to get a raspberry pi image to work on QEMU running in a VBox virtual Ubuntu 20.04 system – my hope was to then try to build mediapipe there much faster than on the “pi” itself; but no luck making that work, too slow.

 

thanks for your help!

jb

**Comment by johncblacker** [comment]
***********update 9/22 & 11:19 ****************

Tried running the build script on my Raspbian lite image and it failed to complete, but differently than on my Raspbian desktop version.

A message was issued about “libass” missing and indicating that an option was require if libass support is not wanted.  I had tried to remove

libass as you suggest, but it wasn’t included on the “lite” version of Raspbian.  I might install it and then uninstall it just to see what happens – maybe

I’ll get a different error.  Anyway, worth a try.

 

jb

**Comment by johncblacker** [comment]
Finally was able to build mediapipe on my rPi.  The short version...installed a clean raspbian stretch and used that to build ffmpeg, but opencv wouldn't build there because of cmake version; upgraded from stretch to buster; built opencv; built mediapipe.  Of course along the way there were dependencies that were missing that I had to resolve, but I didn't want to just install a whole bunch of packages without their being necessary.  Also found a helpful tip on mediapipe github:  here's the link to that post: https://github.com/jiuqiant/mediapipe_python_aarch64/blob/main/README.md
The key part of that post to me was the two "sed" commands.  The fact that it was for aarch64/ rpi4 doesn't matter.  After I was done with the build, I tried it out and had a problem imorting cv2!  Turns out that earlier on in the whole build chain there was a "sudo apt-get install python3-numpy" which was too old, not to mention it took some time before I found out how to get rid of it.  Finally, I got "import cv2" to work and "import mediapipe as mp" to work.  I didn't have time to give it a shot with a real facedection.py script; that's left for today when I get time.  Thanks for your help along the way.  I'm really curious whether I could  successfully build a different version of mediapipe with this sd card setup...
